In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import time
%matplotlib inline

In [2]:
chosen = pd.read_csv('chosen.csv')
chosen.head()

item_cnt_month  mean_month_shop  mean_month_cat  mean_by_year  \
0             1.0         0.034671        0.151543      0.085368   
1             0.0         0.033333        0.151543      0.085368   
2             0.0         0.051541        0.151543      0.085368   
3             0.0         0.050038        0.151543      0.085368   
4             0.0         0.111111        0.151543      0.085368   

   item_cnt_1_ago  item_cnt_2_ago  item_cnt_3_ago  item_cnt_4_ago  
0             1.0             1.0             1.0             1.0  
1             0.0             0.0             0.0             0.0  
2             2.0             2.0             0.0             0.0  
3             0.0             0.0             0.0             0.0  
4             0.0             3.0             1.0             3.0

In [3]:
chosen[chosen.isna().any(axis=1)]

Empty DataFrame
Columns: [item_cnt_month, mean_month_shop, mean_month_cat, mean_by_year, item_cnt_1_ago, item_cnt_2_ago, item_cnt_3_ago, item_cnt_4_ago]
Index: []

In [4]:
chosen['item_cnt_month'].value_counts()

0.000000e+00     8051990
1.000000e+00      897499
2.000000e+00      227553
3.000000e+00       88164
4.000000e+00       45462
5.000000e+00       27138
6.000000e+00       17536
7.000000e+00       11945
2.000000e+01       10705
8.000000e+00        8828
9.000000e+00        6367
1.000000e+01        5104
1.100000e+01        3786
1.200000e+01        3131
1.300000e+01        2572
1.400000e+01        2048
1.500000e+01        1750
1.600000e+01        1460
1.700000e+01        1323
1.800000e+01        1069
1.900000e+01         954
3.315618e-316          1
Name: item_cnt_month, dtype: int64

In [5]:
chosen[(chosen['item_cnt_month'] < 1) & (chosen['item_cnt_month'] > 0)]

item_cnt_month  mean_month_shop  mean_month_cat  mean_by_year  \
1064113   3.315618e-316         0.027951         0.05481      0.085368   

         item_cnt_1_ago  item_cnt_2_ago  item_cnt_3_ago  item_cnt_4_ago  
1064113             0.0             0.0             0.0             0.0

In [6]:
chosen.drop(1064113, axis=0, inplace=True)
chosen['item_cnt_month'].value_counts()

0.0     8051990
1.0      897499
2.0      227553
3.0       88164
4.0       45462
5.0       27138
6.0       17536
7.0       11945
20.0      10705
8.0        8828
9.0        6367
10.0       5104
11.0       3786
12.0       3131
13.0       2572
14.0       2048
15.0       1750
16.0       1460
17.0       1323
18.0       1069
19.0        954
Name: item_cnt_month, dtype: int64

In [7]:
from sklearn.model_selection import StratifiedShuffleSplit
import time
start = time.time()
splitter = StratifiedShuffleSplit(n_splits=1, test_size=0.25, random_state=42)
for train_index, test_index in splitter.split(chosen, chosen['item_cnt_month']):
    X_train = chosen.loc[train_index].drop('item_cnt_month', axis=1)
    y_train = chosen.loc[train_index]['item_cnt_month']
    X_test = chosen.loc[test_index].drop('item_cnt_month', axis=1)
    y_test = chosen.loc[test_index]['item_cnt_month']
# print('Run time: {} minutes.'.format(round((time.time() - start) / 60)))
print('Run time: {} seconds.'.format(time.time() - start))

C:\Users\Chris\Anaconda3\envs\DataScience1\lib\site-packages\ipykernel_launcher.py:6: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  
C:\Users\Chris\Anaconda3\envs\DataScience1\lib\site-packages\ipykernel_launcher.py:7: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  import sys


Run time: 12.874839305877686 seconds.


In [8]:
X_train.shape

(7062288, 7)

In [9]:
y_train.shape

(7062288,)

In [10]:
X_test.shape

(2354096, 7)

In [11]:
y_test.shape

(2354096,)

In [12]:
y_train.value_counts() / y_train.shape[0]

0.0     0.855038
1.0     0.095355
2.0     0.024169
3.0     0.009367
4.0     0.004839
5.0     0.002875
6.0     0.001853
7.0     0.001285
20.0    0.001139
8.0     0.000932
9.0     0.000675
10.0    0.000540
11.0    0.000407
12.0    0.000335
13.0    0.000275
14.0    0.000217
15.0    0.000187
16.0    0.000156
17.0    0.000142
18.0    0.000113
19.0    0.000102
Name: item_cnt_month, dtype: float64

In [13]:
y_test.value_counts() / y_test.shape[0]

0.0     0.855303
1.0     0.095186
2.0     0.024156
3.0     0.009351
4.0     0.004795
5.0     0.002903
6.0     0.001891
7.0     0.001219
20.0    0.001129
8.0     0.000953
9.0     0.000681
10.0    0.000548
11.0    0.000387
12.0    0.000325
13.0    0.000268
14.0    0.000218
15.0    0.000183
16.0    0.000151
17.0    0.000137
18.0    0.000114
19.0    0.000101
Name: item_cnt_month, dtype: float64

In [14]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import GridSearchCV

In [15]:
X_train.max(axis=0)

mean_month_shop     0.697519
mean_month_cat     34.363889
mean_by_year        0.085368
item_cnt_1_ago     20.000000
item_cnt_2_ago     20.000000
item_cnt_3_ago     20.000000
item_cnt_4_ago     20.000000
dtype: float64

In [16]:
y_train.max()

20.0

In [17]:
X_train[X_train.isna().any(axis=1)]

mean_month_shop  mean_month_cat  mean_by_year  item_cnt_1_ago  \
1064113              NaN             NaN           NaN             NaN   

         item_cnt_2_ago  item_cnt_3_ago  item_cnt_4_ago  
1064113             NaN             NaN             NaN

In [18]:
X_train.drop(1064113, axis=0, inplace=True)

In [19]:
X_train[X_train.isna().any(axis=1)]

Empty DataFrame
Columns: [mean_month_shop, mean_month_cat, mean_by_year, item_cnt_1_ago, item_cnt_2_ago, item_cnt_3_ago, item_cnt_4_ago]
Index: []

In [20]:
y_train.drop(1064113, inplace=True)

### Let's combine our training set + validation set into a new full training set to train our final models on.

In [32]:
full_X_train = pd.concat([X_train, X_test])
full_X_train.shape

(9416383, 7)

In [33]:
full_X_train.head()

mean_month_shop  mean_month_cat  mean_by_year  item_cnt_1_ago  \
5777024         0.054834        0.058156      0.069746             0.0   
6030030         0.129154        0.008629      0.069746             0.0   
8245625         0.060713        0.050536      0.050812             0.0   
8222488         0.089250        0.050771      0.050812             0.0   
7385422         0.105202        0.053599      0.050812             0.0   

         item_cnt_2_ago  item_cnt_3_ago  item_cnt_4_ago  
5777024             0.0             0.0             0.0  
6030030             0.0             0.0             0.0  
8245625             0.0             0.0             0.0  
8222488             1.0             1.0             1.0  
7385422             0.0             0.0             0.0

In [34]:
full_X_train.tail()

mean_month_shop  mean_month_cat  mean_by_year  item_cnt_1_ago  \
7817186         0.059209        0.049138      0.050812             0.0   
3400763         0.064862        0.222222      0.069746             0.0   
8875523         0.061961        0.010627      0.050812             0.0   
8458373         0.035799        0.056670      0.050812             0.0   
9345999         0.088287        0.050203      0.050812             0.0   

         item_cnt_2_ago  item_cnt_3_ago  item_cnt_4_ago  
7817186             0.0             1.0             0.0  
3400763             0.0             0.0             0.0  
8875523             0.0             0.0             0.0  
8458373             0.0             0.0             0.0  
9345999             0.0             0.0             0.0

In [35]:
full_y_train = pd.concat([y_train, y_test])
full_y_train.shape

(9416383,)

### Training the models that we'll use on the final test set: Gradient Boosting and Random Forest

The GradientBoostingRegressor (which I'll refer to as GBR) hyperparameters were selected for achieving the best accuracy in the experiments conducted in the Chris_modelling notebook. The RandomForestRegressor (RFR) hyperparameters were selected for achieving the best accuracy in testing performed by my colleague George Ayo. The RFR is included mainly for comparison purposes.

In [36]:
test_gbr = GradientBoostingRegressor(n_estimators=137, learning_rate=0.1, max_depth=11)

In [37]:
start = time.time()
test_gbr.fit(full_X_train, full_y_train)
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))

Run time: 999.3 minutes.


In [38]:
from sklearn.ensemblesemble import RandomForestRegressor
start = time.time()
rfr = RandomForestRegressor(n_estimators=100, max_features=6, n_jobs=3)
rfr.fit(full_X_train, full_y_train)
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))

Run time: 37.3 minutes.


Let's see what the errors are like on the training set for the RFR, and the GBR.

In [21]:
from sklearn.metrics import mean_squared_error

In [39]:
start = time.time()
rfr_full_train_preds = rfr.predict(full_X_train)
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))
mean_squared_error(full_y_train, rfr_full_train_preds)

Run time: 4.2 minutes.


0.5218255017049552

In [40]:
start = time.time()
gbr_full_train_preds = test_gbr.predict(full_X_train)
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))
mean_squared_error(full_y_train, gbr_full_train_preds)

Run time: 1.0 minutes.


0.7313994989088518

The RFR's error on the training set (~0.522) is lower than that of the GBR (~0.731), but this could be because of overfitting.

### Let's check out the feature importances of the RFR and GBR

In [73]:
print('Gradient Boosting Feature Importances')
for i in range(0, 7):
    print(full_X_train.columns[i], ':', test_gbr.feature_importances_[i])

Gradient Boosting Feature Importances
mean_month_shop : 0.06481475085208863
mean_month_cat : 0.12531332793560718
mean_by_year : 0.018613122374106162
item_cnt_1_ago : 0.6699262433703854
item_cnt_2_ago : 0.053451476116158504
item_cnt_3_ago : 0.04239346615117474
item_cnt_4_ago : 0.025487613200479223


In [74]:
print('Random Forest Feature Importances')
for i in range(0, 7):
    print(full_X_train.columns[i], ':', rfr.feature_importances_[i])

Random Forest Feature Importances
mean_month_shop : 0.16149043275566233
mean_month_cat : 0.16889446307543263
mean_by_year : 0.03443288085536867
item_cnt_1_ago : 0.4471653470599825
item_cnt_2_ago : 0.09358217330520499
item_cnt_3_ago : 0.0516367791079396
item_cnt_4_ago : 0.04279792384040929


For the GBR, item_cnt_1_ago is more important than every other feature put together and mean_month_cat is the second-most-important by a large margin. For the RFR, the importances are more spread out but item_cnt_1_ago is still the comparison variable in close to half of all split nodes. However, for the RFR mean_month_shop is more than twice as important as it is for the GBR--almost as important as mean_month_cat.

### Let's train a linear regression that takes the predictions of the RFR and GBR as inputs (model stacking).

In [41]:
start = time.time()
train_preds_df = pd.DataFrame({'rfr': rfr_full_train_preds, 'gbr': gbr_full_train_preds}, columns=['rfr', 'gbr'])
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))
train_preds_df.head()

Run time: 0.0 minutes.


rfr       gbr
0  0.064051  0.075127
1  0.004353  0.039522
2  0.000000  0.075635
3  0.089000  0.271500
4  0.040669  0.094820

In [42]:
train_preds_df.memory_usage(deep=True)

Index          80
rfr      75331064
gbr      75331064
dtype: int64

In [43]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression(n_jobs=3)
start = time.time()
lr.fit(train_preds_df, full_y_train)
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))

Run time: 0.0 minutes.


In [44]:
lr_full_train_preds = lr.predict(train_preds_df)
mean_squared_error(full_y_train, lr_full_train_preds)

0.5063181252471995

Even lower than that of the RFR! Let's hope it's not just really overfitting. Submitting test set predictions to kaggle will tell us.

## Time to load the prepared test data and make predictions on it.

In [45]:
start = time.time()
test_real = pd.read_csv('test_features.csv')
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))
test_real.head()

Run time: 0.0 minutes.


ID  mean_month_shop  mean_month_cat  mean_by_year  item_cnt_1_ago  \
0   0         0.059517        0.192317      0.050812             0.0   
1   1         0.059517        0.075187      0.050812             0.0   
2   2         0.059517        0.192317      0.050812             1.0   
3   3         0.059517        0.181670      0.050812             0.0   
4   4         0.059517        0.471762      0.050812             0.0   

   item_cnt_2_ago  item_cnt_3_ago  item_cnt_4_ago  
0             1.0             3.0             1.0  
1             0.0             0.0             0.0  
2             3.0             1.0             0.0  
3             0.0             1.0             0.0  
4             0.0             0.0             0.0

In [46]:
test_real.shape

(214200, 8)

In [47]:
test_ID = test_real['ID']
test_real.drop('ID', axis=1, inplace=True)

In [48]:
rfr_test_preds = rfr.predict(test_real)
gbr_test_preds = test_gbr.predict(test_real)
test_preds_df = pd.DataFrame({'rfr': rfr_test_preds, 'gbr': gbr_test_preds}, columns=['rfr', 'gbr'])
test_preds_df.head()

rfr       gbr
0  0.740000  0.627867
1  0.110753  0.087372
2  0.626667  0.676305
3  0.010667  0.179143
4  0.020000  0.364703

In [49]:
lr_test_preds = lr.predict(test_preds_df)
lr_test_preds[:5]

array([ 0.78158157,  0.11168131,  0.61250296, -0.05390771, -0.10294255])

Negative predictions will certainly bring our score down, so I better fix those. I'll do that after adding the values to a dataframe (and before I write them to the submission file).

In [50]:
rfr_test_preds_df = pd.DataFrame({'ID': test_ID, 'item_cnt_month': rfr_test_preds}, columns=['ID', 'item_cnt_month'])
rfr_test_preds_df.head()

ID  item_cnt_month
0   0        0.740000
1   1        0.110753
2   2        0.626667
3   3        0.010667
4   4        0.020000

In [51]:
gbr_test_preds_df = pd.DataFrame({'ID': test_ID, 'item_cnt_month': gbr_test_preds}, columns=['ID', 'item_cnt_month'])
gbr_test_preds_df.head()

ID  item_cnt_month
0   0        0.627867
1   1        0.087372
2   2        0.676305
3   3        0.179143
4   4        0.364703

In [52]:
lr_test_preds_df = pd.DataFrame({'ID': test_ID, 'item_cnt_month': lr_test_preds}, columns=['ID', 'item_cnt_month'])
lr_test_preds_df.head()

ID  item_cnt_month
0   0        0.781582
1   1        0.111681
2   2        0.612503
3   3       -0.053908
4   4       -0.102943

In [53]:
# Increase negative predictions to 0
lr_test_preds_df['item_cnt_month'] = lr_test_preds_df['item_cnt_month'].apply(lambda x: x if x >= 0 else 0)
lr_test_preds_df.head()

ID  item_cnt_month
0   0        0.781582
1   1        0.111681
2   2        0.612503
3   3        0.000000
4   4        0.000000

# Save the predictions to CSV files and submit them to Kaggle.

In [54]:
rfr_test_preds_df.to_csv('test_preds_rfr.csv', index=False)

In [55]:
gbr_test_preds_df.to_csv('test_preds_gbr.csv', index=False)

In [56]:
lr_test_preds_df.to_csv('test_preds_lr.csv', index=False)

These predictions from all three of these models was submitted to Kaggle and the RMSE results are:
1.05691 for the RandomForestRegressor,
1.03955 for GradientBoostingRegressor and
1.08329 for the LinearRegression.

Both the RandomForestRegressor and LinearRegression had lower errors on the training set than the GradientBoostingRegressor, and higher errors on the test set, so those two were more overfit. Our best model, the GBR, gives a ranking of 902nd place out of 1929 competitors.

## Saving our models
We do this so that if we want to use them later we won't have to go through the training process again.

In [62]:
from sklearn.externals import joblib

In [65]:
start = time.time()
joblib.dump(test_gbr, 'gbr_windows_sklearn20_0.joblib')
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))

Run time: 0.0 minutes.


gbr_windows_sklearn20_0.pkl ends up being 31,857 KB, which I find surprisingly small. That's because, since the model's max_depth=11, each of the GBR's 137 estimators may have as many 2^11 = 2048 leaf nodes. So there may be as many as 280,576 leaf nodes in the ensemble, each of which has a different output value. Besides all of these output values, the threshold and variable id must be stored for every split node, and there may be as many as 2^10 = 1024 split nodes per estimator, for as many as 140,288 split nodes in total. Now that I calculate it out, assuming 64 bits for each of these numbers, I see the potential total is 35,072 KB for just these numbers. That doesn't include the overhead, such as the structural position of every node in every tree, and whatever model metadata is needed to load the model into memory. Therefore I'm going to try loading the model again and see if it makes the same predictions.

In [66]:
start = time.time()
loaded_gbr = joblib.load('gbr_windows_sklearn20_0.joblib')
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))

Run time: 0.0 minutes.


In [67]:
loaded_gbr

GradientBoostingRegressor(alpha=0.9, criterion='friedman_mse', init=None,
             learning_rate=0.1, loss='ls', max_depth=11, max_features=None,
             max_leaf_nodes=None, min_impurity_decrease=0.0,
             min_impurity_split=None, min_samples_leaf=1,
             min_samples_split=2, min_weight_fraction_leaf=0.0,
             n_estimators=137, n_iter_no_change=None, presort='auto',
             random_state=None, subsample=1.0, tol=0.0001,
             validation_fraction=0.1, verbose=0, warm_start=False)

In [68]:
loaded_gbr_preds = loaded_gbr.predict(test_real)
mean_squared_error(gbr_test_preds, loaded_gbr_preds)

0.0

###### Yup, it seems that the GBR model is reliably stored!

In [70]:
start = time.time()
joblib.dump(rfr, 'rfr_windows_sklearn20_0.pkl')
print('Run time: {} minutes.'.format(round((time.time() - start) / 60, 1)))

Run time: 1.6 minutes.


I assumed saving the RFR would be simpler than saving the GBR because that had more estimators with greater depth. This assumption was incorrect. Saving the RFR to file took 1.6 minutes rather than a couple seconds, and produced a file 7,813,370 KB in size! Almost 8 GB! It must be storing everything about every sample in every leaf node of every tree!